runtime/include/json.h

```diff
@@ -0,0 +1,284 @@
+		fprintf(stderr, "Unable to parse uint8_t for key %s\n", key);
+		return -1;
+	}
+
+	*ret = (uint8_t)temp;
+	return 0;
+}
+
+static inline int
+parse_uint16_t(jsmntok_t tok, const char *json_buf, const char *key, uint16_t *ret)
+{
+	char    *end  = NULL;
+	intmax_t temp = strtoimax(&json_buf[tok.start], &end, 10);
+
+	if (end != &json_buf[tok.end] || temp < 0 || temp > UINT16_MAX) {
+		fprintf(stderr, "Unable to parse uint16_t for key %s\n", key);
+		return -1;
+	}
+
+	*ret = (uint16_t)temp;
+	return 0;
+}
+
+static inline int
+parse_uint32_t(jsmntok_t tok, const char *json_buf, const char *key, uint32_t *ret)
+{
+	char     *end  = NULL;
+	uintmax_t temp = strtoimax(&json_buf[tok.start], &end, 10);
+
+	if (end != &json_buf[tok.end] || temp < 0 || temp > UINT32_MAX) {
+		fprintf(stderr, "Unable to parse uint32_t for key %s\n", key);
+		return -1;
+	}
+
+	*ret = (uint32_t)temp;
+	return 0;
+}
+
+/**
+ * Parses a JSON file into an array of module configs
+ * @param file_name The path of the JSON file
+ * @return module_config_vec_len on success. -1 on Error
+ */
+static inline int
+parse_json(const char *json_buf, ssize_t json_buf_size, struct module_config **module_config_vec)
+{
+	assert(json_buf != NULL);
+	assert(json_buf_size > 0);
+	assert(module_config_vec != NULL);
+
+	jsmntok_t tokens[JSON_MAX_ELEMENT_SIZE * JSON_MAX_ELEMENT_COUNT];
+	int       module_config_vec_len = 0;
+
+
+	/* Initialize the Jasmine Parser and an array to hold the tokens */
+	jsmn_parser module_parser;
+	jsmn_init(&module_parser);
+
+	/* Use Jasmine to parse the JSON */
+	int total_tokens = jsmn_parse(&module_parser, json_buf, json_buf_size, tokens,
+	                              sizeof(tokens) / sizeof(tokens[0]));
+	if (total_tokens < 0) {
+		if (total_tokens == JSMN_ERROR_INVAL) {
+			fprintf(stderr, "Error parsing %s: bad token, JSON string is corrupted\n", json_buf);
+		} else if (total_tokens == JSMN_ERROR_PART) {
+			fprintf(stderr, "Error parsing %s: JSON string is too short, expecting more JSON data\n",
+			        json_buf);
+		} else if (total_tokens == JSMN_ERROR_NOMEM) {
+			/*
+			 * According to the README at https://github.com/zserge/jsmn, this is a potentially recoverable
+			 * error. More tokens can be allocated and jsmn_parse can be re-invoked.
+			 */
+			fprintf(stderr, "Error parsing %s: Not enough tokens, JSON string is too large\n", json_buf);
+		}
+		goto err;
+	}
+
+	if (tokens[0].type != JSMN_ARRAY) {
+		fprintf(stderr, "Outermost Config should be a JSON array, was a JSON %s\n", jsmn_type(tokens[0].type));
+		goto err;
+	}
+
+	module_config_vec_len = tokens[0].size;
+	if (module_config_vec_len == 0) {
+		fprintf(stderr, "Config is an empty JSON array\n");
+		goto err;
+	}
+
+	*module_config_vec = (struct module_config *)calloc(module_config_vec_len, sizeof(struct module_config));
+	int module_idx     = -1;
+	int module_fields_remaining = 0;
+
+	for (int i = 1; i < total_tokens; i++) {
```

Comment:
Why start from `i=1`? This is `jsmn` thing?

Reply:
It's just because we've already checked that the 0th element was a JSON array above this.